onnxruntime/core/graph/graph.cc

```diff
@@ -3753,56 +3794,99 @@ Status Graph::InlineFunction(Node& node) {
   for (auto output_edge : output_edges) {
     RemoveEdge(node.Index(), output_edge.GetNode().Index(), output_edge.GetSrcArgIndex(), output_edge.GetDstArgIndex());
   }
+
+  // Map of function input outputs to nodes input/outputs
   std::unordered_map<std::string, NodeArg*> remap_input_output;
-  if (node.MutableInputDefs().size() != subgraph.GetInputsIncludingInitializers().size())
-    return Status(ONNXRUNTIME, FAIL, "Node " + node.Name() + "'s number of inputs is different from function body graph's number of input.");
+  // Set of node input output names as these names need to be preserved during inlining
+  std::unordered_set<std::string> func_input_output_names;
+
+  ORT_ENFORCE(node.MutableInputDefs().size() == subgraph.GetInputsIncludingInitializers().size(),
+              "Node " + node.Name() + "'s number of inputs is different from function body graph's number of input.");
+
+  ORT_ENFORCE(node.MutableOutputDefs().size() == subgraph.GetOutputs().size(),
+              "Node ", node.Name(), "'s number of outputs is different from function body graph's number of outputs.");
+
   for (size_t i = 0; i < subgraph.GetInputsIncludingInitializers().size(); ++i) {
     auto* input = subgraph.GetInputsIncludingInitializers()[i];
-    if (input->Name() != node.MutableInputDefs()[i]->Name())
+    if (input->Name() != node.MutableInputDefs()[i]->Name()) {
       remap_input_output[input->Name()] = node.MutableInputDefs()[i];
+    }
+    func_input_output_names.insert(input->Name());
   }
 
-  ORT_ENFORCE(node.MutableOutputDefs().size() == subgraph.GetOutputs().size(),
-              "Node ", node.Name(), "'s number of outputs is different from function body graph's number of outputs.");
   for (size_t i = 0; i < subgraph.GetOutputs().size(); ++i) {
     auto* output = subgraph.GetOutputs()[i];
-    if (output->Name() != node.MutableOutputDefs()[i]->Name())
+    if (output->Name() != node.MutableOutputDefs()[i]->Name()) {
       remap_input_output[output->Name()] = node.MutableOutputDefs()[i];
+    }
+    func_input_output_names.insert(output->Name());
   }
 
+  // create a uniq_identifier to append to every node name and intermidiate input\outputs
+  // to make sure there are no unintended duplicates
+  std::stringstream ss;
+  ss << static_cast<const void*>(&node);
+  auto uniq_identifier = ss.str();
+
   RemoveNode(node.Index());
+
   const auto& model_path = ModelPath();
   for (const auto& subgraph_node : subgraph.Nodes()) {
     if (subgraph_node.OpType() == kConstant) {
       // Copy constant nodes _value to name_to_initial_tensor_
```

Comment:
Currently the Graph constructor converts Constant nodes to initializers, so I'm wondering if any Constant nodes would actually exist at this point in time.

Reply:
yes they do exist inside the function body

Reply:
`subgraph` is an ORT Graph instance isn't it, so subgraph_node is an ORT Node? as the ORT Graph loading converts Constant nodes to initializers, afaik this is impossible. if you were dealing with nodes at the proto level it would be.

Reply:
My bad for not adding detailed comments in my previous response. 

Function body initialization follows a bit different flow. 

The [graph is constructed first](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/graph/function.cc#L283) and then the [nodes are added](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/graph/function.cc#L473) to the graph. This is why in the graph constructor [at this point](https://github.com/microsoft/onnxruntime/blob/53eb79f9f6715d2af43dc68abcdde798565a1d69/onnxruntime/core/graph/graph.cc#L1028-L1036) there are no nodes in the graph and therefore no const node-> initializer conversion happens

Because of the above mentioned reason an initialized function body does have constant nodes and the conversion to initializers happens during function inlining.

Reply:
Would it be more consistent to change AddNode to convert a Constant node into an initializer? Maybe as a separate PR. Seems overly complicated to have to deal with edge cases like this.